README
NAME
    Email::Simple - Simple parsing of RFC2822 message format and headers

SYNOPSIS
        my $mail = Email::Simple->new($text);

        my $from_header = $mail->header("From");
        my @received = $mail->header("Received");

        $mail->header_set("From", 'Simon Cozens <simon@cpan.org>');

        my $old_body = $mail->body;
        $mail->body_set("Hello world\nSimon");

        print $mail->as_string;

        # AND THAT'S ALL.

DESCRIPTION
    "Email::Simple" is the first deliverable of the "Perl Email Project", a
    reaction against the complexity and increasing bugginess of the
    "Mail::*" modules. In contrast, "Email::*" modules are meant to be
    simple to use and to maintain, pared to the bone, fast, minimal in their
    external dependencies, and correct.

        Can you sum up plan 9 in layman's terms?
        It does everything Unix does only less reliably - kt

METHODS
    Methods are deliberately kept to a minimum. This is meant to be simple.
    No, I will not add method X. This is meant to be simple. Why doesn't it
    have feature Y? Because it's meant to be simple.

  new
    Parse an email from a scalar containing an RFC2822 formatted message,
    and return an object.

  header
    Returns a list of the contents of the given header.

    If called in scalar context, will return the first header so named. I'm
    not sure I like that. Maybe it should always return a list. But it
    doesn't.

  header_set
        $mail->header_set($field, $line1, $line2, ...);

    Sets the header to contain the given data. If you pass multiple lines
    in, you get multiple headers, and order is retained.

  body
    Returns the body text of the mail.

  body_set
    Sets the body text of the mail.

  as_string
    Returns the mail as a string, reconstructing the headers. Please note
    that header fields are kept in order if they are unique, but, for,
    instance, multiple "Received" headers will be grouped together. (This is
    in accordance with RFC2822, honest.)

    Also, if you've added new headers with "header_set" that weren't in the
    original mail, they'll be added to the end.

CAVEATS
    Email::Simple handles only RFC2822 formatted messages. This means you
    cannot expect it to cope well as the only parser between you and the
    outside world, say for example when writing a mail filter for invocation
    from a .forward file (for this we recommend you use Email::Filter
    anyway). For more information on this issue please consult RT issue
    2478, http://rt.cpan.org/NoAuth/Bug.html?id=2478 .

PERL EMAIL PROJECT

This module is maintained by the Perl Email Project

  http://emailproject.perl.org/wiki/Email::Simple

COPYRIGHT AND LICENSE
    Copyright 2004 by Casey West

    Copyright 2003 by Simon Cozens

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl extension Email::Simple

{{$NEXT}}

2.218     2023-01-08 19:49:09-05:00 America/New_York
        - update author contact info
        - bump version required to v5.12.0 (it was already effectively that
          after some upstream changes)

2.217     2020-11-02 19:13:16-05:00 America/New_York (TRIAL RELEASE)
        - add ->header_rename to header object
        - issue a warning on non-ASCII codepoints added to message (thanks,
          Pali Rohár)

2.216     2018-06-06 11:47:52-04:00 America/New_York
        - no changes since v2.215

2.215     2018-05-04 16:21:28+10:00 Australia/Melbourne (TRIAL RELEASE)
        - do not re-fold folded lines (Marc Bradshaw)

2.214     2017-09-08 14:21:20-04:00 America/New_York
        - the method header_raw_prepend is now available on emails and headers

2.213     2017-01-11 09:21:54-05:00 America/New_York
        - no changes since v2.212

2.212     2016-12-18 20:05:09-05:00 America/New_York (TRIAL RELEASE)
        - add an $index arg to ->header to get the nth header (thanks, Pali
          Rohár)

2.211     2016-11-12 09:39:09-05:00 America/New_York
        - fix a fencepost error in headers and header_names that led to a
          spurious undef in return and to a warning (thanks, Pali Rohár)

2.210     2016-03-05 12:16:16-05:00 America/New_York
        - fix a bug where folding treated the string "0" as an empty string
          (thanks, Pali Rohár)

2.209     2015-10-11 22:57:04-04:00 America/New_York (TRIAL RELEASE)
        - avoid opportunity for pathological backtracking behavior on
          malformed messages

2.208     2015-07-20 20:34:04-04:00 America/New_York
        - same as v2.207
        - ...but bump perl required to v5.8

2.207     2015-07-04 14:23:36-04:00 America/New_York (TRIAL RELEASE)
        - fold headers passed to header arg in create method
          https://github.com/rjbs/Email-Simple/issues/7

2.206     2015-03-26 23:30:51-04:00 America/New_York
        - the changes from 2.204 are back (and re-listed here), but now with
          the added benefit of not breaking Email::MIME; see the git repository
          for the sordid details
        - provide header_raw and header_raw_set for consistent behavior between
          Email::Simple and Email::MIME; also, header_raw_pairs

2.205     2015-03-26 14:54:45-04:00 America/New_York
        - undo changes from 2.204 until safe Email::MIME upgrade path is
          available

2.204     2015-03-25 22:50:49-04:00 America/New_York
        - provide header_raw and header_raw_set for consistent behavior between
          Email::Simple and Email::MIME; also, header_raw_pairs

2.203     2014-02-14 21:56:14-05:00 America/New_York
        - translate vertical whitespace in headers into horizontal whitespace,
          with a warning

2.202     2013-09-04 09:48:33 America/New_York
          repackage with Dist::Zilla; update repo and bug tracker

2.201     2013-04-16
          no changes

2.200_01  2013-04-09
          preserve the original header layout unless changed

2.103     2013-04-09
          fix a bug in header unfolding where "0\n 1" could be unfolded to "01"

2.102     2012-07-13
          recognize crlf correctly even if there is no body (thanks, Christian
          Walde)

2.101     2011-12-23
          remove instruction to use ::Creator instead of just ->create

          documentation improvements (thanks, Michael Stevens)

2.100     2009-11-03
          merge Email-Simple-Creator into Email-Simple distribution

2.005     2009-01-22
          add repository metadata

2.004     2008-06-25
          publicize default_header_class method
          remove the long-deprecated Headers.pm

2.003     2007-07-19
          improve tests, for 5.5 compat
          remove conditional circular prereq on Email::MIME, replace it with a
          warning and delay

2.002     2007-07-14
          change initialization order to unbreak Email::MIME
          do not return ->body from ->body_set to simplify subclass behavior

2.001     2007-07-13
          fix t/perl-minver.t to properly skip if T::MV not installed
          retroactively set 2.000 release date!

2.000     2007-07-13
          huge improvement to speed of bodyless message parsing
            pointed out by Dan Dascalescu; thanks!
          more documentation of header class
          new Header crlf defaults to real CRLF
          fix tests to avoid requiring 5.6

1.999     2007-03-20
          fix bug 25496: deletion of headers affected the wrong range,
            sometimes deleting too many headers -- thanks, Nicholas Oxhoej!
          fix bug 24922: errant space in last header of CRLF-delim email
            thanks, Barry Downes and Alex Vandiver

1.998     2007-02-07
          MAJOR REFACTORING OF GUTS
          If you run Email::MIME, you MUST be running Email::MIME 1.857 or
            better.
          require Email::MIME 1.857 in Makefile.PL only if an older
            version is already installed
          boldly moving forward with refactored headers and
            Email::Simple:::Header
          greatly reduce memory footprint
          add crlf method to allow other modules to avoid ->{mycrlf}
          fix broken header-junk test

1.996    2006-11-27
  - do not wrap Content-Type field; it can cause Outlook to go nuts
    while this is a special case, it does not make E::S incorrect, as folding
    long headers is allowed (for aesthetics) and not mandated

1.995    2006-10-19
  - tentative refactoring of headers

1.992    2006-10-05

  - fix a number of bugs when setting multiple headers, which would often
    refuse to set more values than were currently present
  - added a test for prepending (rather than appending) headers; while E::S
    does not yet support this, the header behavior will be easier to replace in
    future versions, and this is a forward-looking test

1.990    2006-09-05

  - ->header('foo') returns false if there is no foo header (formerly '')

Makefile.PL
use strict;
use ExtUtils::MakeMaker;

use 5.00503;

# This is so stupid!  We need to make sure that Email::MIME, a downstream
# module, is running a version that doesn't screw around with the guts of
# Email::Simple.
if (eval { require Email::MIME }) {
  unless (eval { Email::MIME->VERSION(1.857) }) {
    warn <<END_ACHTUNG;
### ACHTUNG!  You need to update Email::MIME to a later version, as versions
### before 1.857 meddled in the guts of Email::Simple, which have been changed.
### If you are using an automated installer, this should happen automatically.
###
### This installation will proceed in five seconds.
END_ACHTUNG
    sleep 5;
  }
}

WriteMakefile(
  NAME          => 'Email::Simple',
  VERSION_FROM  => 'lib/Email/Simple.pm',
  (eval { ExtUtils::MakeMaker->VERSION(6.21) } ? (LICENSE => 'perl') : ()),
  PREREQ_PM     => {
    'Test::More' => '0.47',
    'Email::Date::Format' => 0,
  },
  (eval { ExtUtils::MakeMaker->VERSION(6.46) }
    ? (META_MERGE => {
        resources => {
          repository => 'http://github.com/rjbs/email-simple'
        }
      })
    : ()
  ),
);


lib/Email/Simple/Header.pm
use v5.12.0;
use warnings;
package Email::Simple::Header;
# ABSTRACT: the header of an Email::Simple message

use Carp ();

our @CARP_NOT = qw(Email::Simple);

require Email::Simple;

=head1 SYNOPSIS

  my $email = Email::Simple->new($text);

  my $header = $email->header_obj;
  print $header->as_string;

=head1 DESCRIPTION

This method implements the headers of an Email::Simple object.  It is a very
minimal interface, and is mostly for private consumption at the moment.

=method new

  my $header = Email::Simple::Header->new($head, \%arg);

C<$head> is a string containing a valid email header, or a reference to such a
string.  If a reference is passed in, don't expect that it won't be altered.

Valid arguments are:

  crlf - the header's newline; defaults to CRLF

=cut

# We need to be able to:
#   * get all values by lc name
#   * produce all pairs, with case intact

sub new {
  my ($class, $head, $arg) = @_;

  my $head_ref = ref $head ? $head : \$head;

  my $self = { mycrlf => $arg->{crlf} || "\x0d\x0a", };

  my $headers = $class->_header_to_list($head_ref, $self->{mycrlf});

  #  for my $header (@$headers) {
  #    push @{ $self->{order} }, $header->[0];
  #    push @{ $self->{head}{ $header->[0] } }, $header->[1];
  #  }
  #
  #  $self->{header_names} = { map { lc $_ => $_ } keys %{ $self->{head} } };
  $self->{headers} = $headers;

  bless $self => $class;
}

sub _header_to_list {
  my ($self, $head, $mycrlf) = @_;

  Carp::carp 'Header with wide characters' if ${$head} =~ /[^\x00-\xFF]/;

  my @headers;

  my $crlf = Email::Simple->__crlf_re;

  while ($$head =~ m/\G(.+?)$crlf/go) {
    local $_ = $1;

    if (/^\s+/ or not /^([^:]+):\s*(.*)/) {
      # This is a continuation line. We fold it onto the end of
      # the previous header.
      next if !@headers;  # Well, that sucks.  We're continuing nothing?

      (my $trimmed = $_) =~ s/^\s+//;
      $headers[-1][0] .= $headers[-1][0] =~ /\S/ ? " $trimmed" : $trimmed;
      $headers[-1][1] .= "$mycrlf$_";
    } else {
      push @headers, $1, [ $2, $_ ];
    }
  }

  return \@headers;
}

=method as_string

  my $string = $header->as_string(\%arg);

This returns a stringified version of the header.

=cut

# RFC 2822, 3.6:
# ...for the purposes of this standard, header fields SHOULD NOT be reordered
# when a message is transported or transformed.  More importantly, the trace
# header fields and resent header fields MUST NOT be reordered, and SHOULD be
# kept in blocks prepended to the message.

sub as_string {
  my ($self, $arg) = @_;
  $arg ||= {};

  my $header_str = '';

  my $headers = $self->{headers};

  my $fold_arg = {
    # at     => (exists $arg->{fold_at} ? $arg->{fold_at} : $self->default_fold_at),
    # indent => (exists $arg->{fold_indent} ? $arg->{fold_indent} : $self->default_fold_indent),
    at     => $self->_default_fold_at,
    indent => $self->_default_fold_indent,
  };

  for (my $i = 0; $i < @$headers; $i += 2) {
    if (ref $headers->[ $i + 1 ]) {
      $header_str .= $headers->[ $i + 1 ][1] . $self->crlf;
    } else {
      my $header = "$headers->[$i]: $headers->[$i + 1]";

      $header_str .= $self->_fold($header, $fold_arg);
    }
  }

  return $header_str;
}

=method header_names

This method returns a list of the unique header names found in this header, in
no particular order.

=cut

sub header_names {
  my $headers = $_[0]->{headers};

  my %seen;
  grep  { !$seen{ lc $_ }++ }
    map { $headers->[ $_ * 2 ] } 0 .. @$headers / 2 - 1;
}

=method header_raw_pairs

  my @pairs = $header->header_raw_pairs;
  my $first_name  = $pairs[0];
  my $first_value = $pairs[1];

This method returns a list of all the field/value pairs in the header, in the
order that they appear in the header.  (Remember: don't try assigning that to a
hash.  Some fields may appear more than once!)

=method header_pairs

L<header_pairs> is another name for L<header_raw_pairs>, which was the original
name for the method and which you'll see most often.  In general, though, it's
better to be explicit and use L<header_raw_pairs>.  (In Email::MIME,
L<header_str_pairs> exists for letting the library do the header decoding for
you.)

=cut

sub header_raw_pairs {
  my ($self) = @_;

  my @pairs = map {; _str_value($_) } @{ $self->{headers} };

  return @pairs;
}

sub header_pairs {
  my ($self) = @_;
  $self->header_raw_pairs;
}

=method header_raw

  my $first_value = $header->header_raw($field);
  my $nth_value   = $header->header_raw($field, $index);
  my @all_values  = $header->header_raw($field);

This method returns the value or values of the given header field.  If the
named field does not appear in the header, this method returns false.

=method header

This method just calls C<header_raw>.  It's the older name for C<header_raw>,
but it can be a problem because L<Email::MIME>, a subclass of Email::Simple,
makes C<header> return the header's decoded value.

=cut

sub _str_value { return ref $_[0] ? $_[0][0] : $_[0] }

sub header_raw {
  my ($self, $field, $index) = @_;

  my $headers  = $self->{headers};
  my $lc_field = lc $field;

  if (wantarray and not defined $index) {
    return map { _str_value($headers->[ $_ * 2 + 1 ]) }
      grep { lc $headers->[ $_ * 2 ] eq $lc_field } 0 .. @$headers / 2 - 1;
  } else {
    $index = 0 unless defined $index;
    my $max = @$headers / 2 - 1;
    my @indexes = $index >= 0 ? (0 .. $max) : reverse(0 .. $max);
    $index = -1-$index if $index < 0;
    for (@indexes) {
      next unless lc $headers->[ $_ * 2 ] eq $lc_field;
      return _str_value($headers->[ $_ * 2 + 1 ]) if $index-- == 0;
    }
    return undef;
  }
}

*header = \&header_raw;

=method header_raw_set

  $header->header_raw_set($field => @values);

This method updates the value of the given header.  Existing headers have their
values set in place.  Additional headers are added at the end.  If no values
are given to set, the header will be removed from to the message entirely.

=method header_set

L<header_set> is another name for L<header_raw_set>, which was the original
name for the method and which you'll see most often.  In general, though, it's
better to be explicit and use L<header_raw_set>.  (In Email::MIME,
L<header_str_set> exists for letting the library do the header encoding for
you.)

=cut

# Header fields are lines composed of a field name, followed by a colon (":"),
# followed by a field body, and terminated by CRLF.  A field name MUST be
# composed of printable US-ASCII characters (i.e., characters that have values
# between 33 and 126, inclusive), except colon.  A field body may be composed
# of any US-ASCII characters, except for CR and LF.

# However, a field body may contain CRLF when used in header "folding" and
# "unfolding" as described in section 2.2.3.

sub header_raw_set {
  my ($self, $field, @data) = @_;

  Carp::carp "Header name '$field' with wide characters" if $field =~ /[^\x00-\xFF]/;
  Carp::carp "Value for '$field' header with wide characters" if grep /[^\x00-\xFF]/, @data;

  # I hate this block. -- rjbs, 2006-10-06
  if ($Email::Simple::GROUCHY) {
    Carp::croak "field name contains illegal characters"
      unless $field =~ /^[\x21-\x39\x3b-\x7e]+$/;
    Carp::carp "field name is not limited to hyphens and alphanumerics"
      unless $field =~ /^[\w-]+$/;
  }

  my $headers = $self->{headers};

  my $lc_field = lc $field;
  my @indices = grep { lc $headers->[$_] eq $lc_field }
    map { $_ * 2 } 0 .. @$headers / 2 - 1;

  if (@indices > @data) {
    my $overage = @indices - @data;
    splice @{$headers}, $_, 2 for reverse @indices[ -$overage .. -1 ];
    pop @indices for (1 .. $overage);
  } elsif (@data > @indices) {
    my $underage = @data - @indices;
    for (1 .. $underage) {
      push @$headers, $field, undef;  # temporary value
      push @indices, $#$headers - 1;
    }
  }

  for (0 .. $#indices) {
    $headers->[ $indices[$_] + 1 ] = $data[$_];
  }

  return wantarray ? @data : $data[0];
}

sub header_set {
  my ($self, $field, @data) = @_;
  $self->header_raw_set($field, @data);
}

=method header_raw_prepend

  $header->header_raw_prepend($field => $value);

This method adds a new instance of the name field as the first field in the
header.

=cut

sub header_raw_prepend {
  my ($self, $field, $value) = @_;

  Carp::confess("tried to prepend raw header with undefined field name")
    unless defined $field;

  Carp::confess(qq{tried to prepend raw header "$field" with undefined value})
    unless defined $value;

  unshift @{ $self->{headers} }, $field => $value;

  return;
}

=method header_rename

  $header->header_rename($field, $new_name, $nth);

This renames the named field to the new name.  If C<$nth> is given, only the
I<n>th instance of the field will be renamed.  It is fatal to rename an
instance that does not exist.  The first instance of a header is the 0th.

If C<$nth> is omitted, all instances of the header are renamed.

When picking headers to rename, C<$field> is matched case insensitively.  So,
given this header:

    happythoughts: yes
    HappyThoughts: so many
    hapPyThouGhts: forever

Then this code...

    $header->rename_header('happythoughts', 'Delights');

...will result in this:

    Delights: yes
    Delights: so many
    Delights: forever

Headers may be rewrapped as a result of renaming.

=cut

sub header_rename {
  my ($self, $field, $new_name, $n) = @_;

  my $headers = $self->{headers};
  my $lc_field = lc $field;

  my @indices = grep { lc $headers->[$_] eq $lc_field }
    map { $_ * 2 } 0 .. @$headers / 2 - 1;

  if (defined $n) {
    if ($n < 0) { Carp::confess("negative header index makes no sense") }
    if ($n > $#indices) { Carp::confess("$n exceeds count of $field headers") }

    @indices = $indices[$n];
  }

  for my $i (@indices) {
    $headers->[$i] = $new_name;
    if (ref $headers->[$i + 1]) {
      $headers->[$i + 1] = $headers->[ $i + 1 ][0];
    }
  }

  return;
}

=method crlf

This method returns the newline string used in the header.

=cut

sub crlf { $_[0]->{mycrlf} }

# =method fold
#
#   my $folded = $header->fold($line, \%arg);
#
# Given a header string, this method returns a folded version, if the string is
# long enough to warrant folding.  This method is used internally.
#
# Valid arguments are:
#
#   at      - fold lines to be no longer than this length, if possible
#             if given and false, never fold headers
#   indent  - indent lines with this string

# =cut

sub _fold {
  my ($self, $line, $arg) = @_;
  $arg ||= {};

  $arg->{at} = $self->_default_fold_at unless exists $arg->{at};

  $arg->{indent} = $self->_default_fold_indent unless exists $arg->{indent};

  my $indent = $arg->{indent} || $self->_default_fold_indent;

  # We will not folder headers if...
  # * the header has vertical whitespace
  # * all vertical whitespace is followed by horizontal whitespace or END
  if ($line =~ /\n/) {
    if ($line =~ s/\n([^\s\t])/\n$indent$1/g) {
      Carp::carp("bad space in header: newline followed by non-space: $line");
    } else {
      $line .= $self->crlf unless $line =~ /\n$/;
      return $line;
    }
  }

  return $line . $self->crlf unless $arg->{at} and $arg->{at} > 0;

  my $limit  = ($arg->{at} || $self->_default_fold_at) - 1;

  return $line . $self->crlf if length $line <= $limit;

  return $self->__fold_objless($line, $limit, $indent, $self->crlf);

}

sub __fold_objless {
  my ($self, $line, $limit, $indent, $crlf) = @_;

  # We know it will not contain any new lines at present
  my $folded = "";
  while (length $line) {
    if ($line =~ s/^(.{0,$limit})(\s|\z)//) {
      $folded .= $1 . $crlf;
      $folded .= $indent if length $line;
    } else {
      # Basically nothing we can do. :(
      $folded .= $line . $crlf;
      last;
    }
  }

  return $folded;
}

# =method default_fold_at
#
# This method (provided for subclassing) returns the default length at which to
# try to fold header lines.  The default default is 78.
#
# =cut

sub _default_fold_at { 78 }

# =method default_fold_indent
#
# This method (provided for subclassing) returns the default string used to
# indent folded headers.  The default default is a single space.
#
# =cut

sub _default_fold_indent { " " }

1;


lib/Email/Simple/Creator.pm
use v5.12.0;
use warnings;
package Email::Simple::Creator;
# ABSTRACT: private helper for building Email::Simple objects

use Carp ();

sub _crlf {
  "\x0d\x0a";
}

sub _date_header {
  require Email::Date::Format;
  Email::Date::Format::email_date();
}

our @CARP_NOT = qw(Email::Simple Email::MIME);

sub _add_to_header {
  my ($class, $header, $key, $value) = @_;
  $value //= '';

  Carp::carp "Header name '$key' with wide characters" if $key =~ /[^\x00-\xFF]/;
  Carp::carp "Value '$value' for '$key' header with wide characters" if $value =~ /[^\x00-\xFF]/;

  if ($value =~ s/[\x0a\x0b\x0c\x0d\x85\x{2028}\x{2029}]+/ /g) {
    Carp::carp("replaced vertical whitespace in $key header with space; this will become fatal in a future version");
  }

  $$header .= Email::Simple::Header->__fold_objless("$key: $value", 78, q{ }, $class->_crlf);
}

sub _finalize_header {
  my ($class, $header) = @_;
  $$header .= $class->_crlf;
}

1;


lib/Email/Simple.pm
use v5.12.0;
use warnings;
package Email::Simple;
# ABSTRACT: simple parsing of RFC2822 message format and headers

use Carp ();

use Email::Simple::Creator;
use Email::Simple::Header;

our $GROUCHY = 0;

# We are liberal in what we accept.
sub __crlf_re { qr/\x0a\x0d|\x0d\x0a|\x0a|\x0d/; }

=head1 SYNOPSIS

  use Email::Simple;
  my $email = Email::Simple->new($text);

  my $from_header = $email->header("From");
  my @received = $email->header("Received");

  $email->header_set("From", 'Simon Cozens <simon@cpan.org>');

  my $old_body = $email->body;
  $email->body_set("Hello world\nSimon");

  print $email->as_string;

...or, to create a message from scratch...

  my $email = Email::Simple->create(
      header => [
        From    => 'casey@geeknest.com',
        To      => 'drain@example.com',
        Subject => 'Message in a bottle',
      ],
      body => '...',
  );

  $email->header_set( 'X-Content-Container' => 'bottle/glass' );

  print $email->as_string;

=head1 DESCRIPTION

The Email:: namespace was begun as a reaction against the increasing complexity
and bugginess of Perl's existing email modules.  C<Email::*> modules are meant
to be simple to use and to maintain, pared to the bone, fast, minimal in their
external dependencies, and correct.

=method new

  my $email = Email::Simple->new($message, \%arg);

This method parses an email from a scalar containing an RFC2822 formatted
message and returns an object.  C<$message> may be a reference to a message
string, in which case the string will be altered in place.  This can result in
significant memory savings.

If you want to create a message from scratch, you should use the C<L</create>>
method.

Valid arguments are:

  header_class - the class used to create new header objects
                 The named module is not 'require'-ed by Email::Simple!

=cut

sub new {
  my ($class, $text, $arg) = @_;
  $arg ||= {};

  Carp::croak 'Unable to parse undefined message' if ! defined $text;

  my $text_ref = (ref $text || '' eq 'SCALAR') ? $text : \$text;

  Carp::carp 'Message with wide characters' if ${$text_ref} =~ /[^\x00-\xFF]/;

  my ($pos, $mycrlf) = $class->_split_head_from_body($text_ref);

  my $self = bless { mycrlf => $mycrlf } => $class;

  my $head;
  if (defined $pos) {
    $head = substr $$text_ref, 0, $pos, '';
    substr($head, -(length $mycrlf)) = '';
  } else {
    $head     = $$text_ref;
    $text_ref = \'';
  }

  my $header_class = $arg->{header_class} || $self->default_header_class;

  $self->header_obj_set(
    $header_class->new(\$head, { crlf => $self->crlf })
  );

  $self->body_set($text_ref);

  return $self;
}

# Given the text of an email, return ($pos, $crlf) where $pos is the position
# at which the body text begins and $crlf is the type of newline used in the
# message.
sub _split_head_from_body {
  my ($self, $text_ref) = @_;

  # For body/header division, see RFC 2822, section 2.1
  #
  # Honestly, are we *ever* going to have LFCR messages?? -- rjbs, 2015-10-11
  my $re = qr{\x0a\x0d\x0a\x0d|\x0d\x0a\x0d\x0a|\x0d\x0d|\x0a\x0a};

  if ($$text_ref =~ /($re)/gsm) {
    my $crlf = substr $1, 0, length($1)/2;
    return (pos($$text_ref), $crlf);
  } else {

    # The body is, of course, optional.
    my $re = $self->__crlf_re;
    $$text_ref =~ /($re)/gsm;
    return (undef, ($1 || "\n"));
  }
}

=method create

  my $email = Email::Simple->create(header => [ @headers ], body => '...');

This method is a constructor that creates an Email::Simple object
from a set of named parameters. The C<header> parameter's value is a
list reference containing a set of headers to be created. The C<body>
parameter's value is a scalar value holding the contents of the message
body.  Line endings in the body will normalized to CRLF.

If no C<Date> header is specified, one will be provided for you based on the
C<gmtime> of the local machine. This is because the C<Date> field is a required
header and is a pain in the neck to create manually for every message. The
C<From> field is also a required header, but it is I<not> provided for you.

=cut

our $CREATOR = 'Email::Simple::Creator';

sub create {
  my ($class, %args) = @_;

  # We default it in here as well as below because by having it here, then we
  # know that if there are no other headers, we'll get the proper CRLF.
  # Otherwise, we get a message with incorrect CRLF. -- rjbs, 2007-07-13
  my $headers = $args{header} || [ Date => $CREATOR->_date_header ];
  my $body    = $args{body} || '';

  my $empty   = q{};
  my $header  = \$empty;

  for my $idx (map { $_ * 2 } 0 .. @$headers / 2 - 1) {
    my ($key, $value) = @$headers[ $idx, $idx + 1 ];
    $CREATOR->_add_to_header($header, $key, $value);
  }

  $CREATOR->_finalize_header($header);

  my $email = $class->new($header);

  $email->header_raw_set(Date => $CREATOR->_date_header)
    unless defined $email->header_raw('Date');

  $body = (join $CREATOR->_crlf, split /\x0d\x0a|\x0a\x0d|\x0a|\x0d/, $body)
        . $CREATOR->_crlf;

  $email->body_set($body);

  return $email;
}


=method header_obj

  my $header = $email->header_obj;

This method returns the object representing the email's header.  For the
interface for this object, see L<Email::Simple::Header>.

=cut

sub header_obj {
  my ($self) = @_;
  return $self->{header};
}

# Probably needs to exist in perpetuity for modules released during the "__head
# is tentative" phase, until we have a way to force modules below us on the
# dependency tree to upgrade.  i.e., never and/or in Perl 6 -- rjbs, 2006-11-28
BEGIN { *__head = \&header_obj }

=method header_obj_set

  $email->header_obj_set($new_header_obj);

This method substitutes the given new header object for the email's existing
header object.

=cut

sub header_obj_set {
  my ($self, $obj) = @_;
  $self->{header} = $obj;
}

=method header

  my @values = $email->header($header_name);
  my $first  = $email->header($header_name);
  my $value  = $email->header($header_name, $index);

In list context, this returns every value for the named header.  In scalar
context, it returns the I<first> value for the named header.  If second
parameter is specified then instead I<first> value it returns value at
position C<$index> (negative C<$index> is from the end).

=method header_set

    $email->header_set($field, $line1, $line2, ...);

Sets the header to contain the given data. If you pass multiple lines
in, you get multiple headers, and order is retained.  If no values are given to
set, the header will be removed from to the message entirely.

=method header_raw

This is another name (and the preferred one) for C<header>.

=method header_raw_set

This is another name (and the preferred one) for C<header_set>.

=method header_raw_prepend

  $email->header_raw_prepend($field => $value);

This method adds a new instance of the name field as the first field in the
header.

=method header_names

    my @header_names = $email->header_names;

This method returns the list of header names currently in the email object.
These names can be passed to the C<header> method one-at-a-time to get header
values. You are guaranteed to get a set of headers that are unique. You are not
guaranteed to get the headers in any order at all.

For backwards compatibility, this method can also be called as B<headers>.

=method header_pairs

  my @headers = $email->header_pairs;

This method returns a list of pairs describing the contents of the header.
Every other value, starting with and including zeroth, is a header name and the
value following it is the header value.

=method header_raw_pairs

This is another name (and the preferred one) for C<header_pairs>.

=cut

BEGIN {
  no strict 'refs';
  for my $method (qw(
    header_raw header
    header_raw_set header_set
    header_raw_prepend
    header_raw_pairs header_pairs
    header_names
  )) {
    *$method = sub { (shift)->header_obj->$method(@_) };
  }
  *headers = \&header_names;
}

=method body

Returns the body text of the mail.

=cut

sub body {
  my ($self) = @_;
  return (defined ${ $self->{body} }) ? ${ $self->{body} } : '';
}

=method body_set

Sets the body text of the mail.

=cut

sub body_set {
  my ($self, $text) = @_;
  my $text_ref = ref $text ? $text : \$text;
  Carp::carp 'Body with wide characters' if defined ${$text_ref} and ${$text_ref} =~ /[^\x00-\xFF]/;
  $self->{body} = $text_ref;
  return;
}

=method as_string

Returns the mail as a string, reconstructing the headers.

=cut

sub as_string {
  my $self = shift;
  return $self->header_obj->as_string . $self->crlf . $self->body;
}

=method crlf

This method returns the type of newline used in the email.  It is an accessor
only.

=cut

sub crlf { $_[0]->{mycrlf} }

=method default_header_class

This returns the class used, by default, for header objects, and is provided
for subclassing.  The default default is Email::Simple::Header.

=cut

sub default_header_class { 'Email::Simple::Header' }

1;

__END__

=head1 CAVEATS

Email::Simple handles only RFC2822 formatted messages.  This means you cannot
expect it to cope well as the only parser between you and the outside world,
say for example when writing a mail filter for invocation from a .forward file
(for this we recommend you use L<Email::Filter> anyway).

=cut
